crates/sui-core/src/state_accumulator/mod.rs

```diff
@@ -0,0 +1,394 @@
+    async fn run(&mut self) {
+        let mut pending: AccumulatorTaskBuffer = FuturesOrdered::new();
+
+        loop {
+            tokio::select! {
+                // process completed tasks
+                Some(Ok(Ok((acc, state)))) = pending.next() => {
+                    // Save checkpoint hash if not already exists
+                    if !self.store.tables.state_hash_by_checkpoint.contains_key(&state.checkpoint_seq_num).unwrap() {
+                        self.store.tables.state_hash_by_checkpoint
+                            .insert(&state.checkpoint_seq_num, &acc)
+                            .expect("StateAccumulator: failed to insert state hash");
+                        self.store.checkpoint_state_notify_read.notify(&state.checkpoint_seq_num, &acc);
+                    }
+
+                    if state.end_of_epoch_flag {
+                        // TODOS:
+                        //
+                        // 1. Here we call this once at the end of the epoch. Depending
+                        // on the performance here, we may instead move to running this as
+                        // a separate background task that accumulates on an ongoing basis.
+                        //
+                        // 2. Here was assume for simplicity that the end of epoch checkpoint will always
+                        // be scheduled last. This is true for now, but we should not assume this.
+                        // Can be fixed to spawn a separate process that does a best effort root digest
+                        // accumulation, with retries in case there are missing checkpoint accumulators.
+                        self.accumulate(state)
+                            .await
+                            .expect("Accumulation failed");
+                    }
+                }
+                // schedule new tasks
+                state = self.queue.recv() => match state {
+                    Some(state) => {
+                        debug!(
+                            ?state,
+                            "StateAccumulator: received enqueued state",
+                        );
+
+                        let store = self.store.clone();
+                        let state_clone = state.clone();
+
+                        pending.push_back(spawn_monitored_task!(async move {
+                            let accumulator = hash_checkpoint_effects(state_clone.clone(), store)?;
+                            Ok((accumulator, state_clone))
+                        }));
+                    }
+                    None => panic!("StateAccumulator: all senders have been dropped"),
+                }
+            }
+        }
+    }
+
+    /// Unions all checkpoint accumulators at the end of the epoch to generate the
+    /// root state hash and saves it. This function is guaranteed to be idempotent (despite the
+    /// underlying data structure not being) as long as it is not called in a multi-threaded
```

Comment:
maybe i've misunderstood, but i think the tables are in fact idempotent and this can be called willy nilly from any number of threads. If we read the accumulated state hash as of some checkpoint C, and then accumulate all checkpoints > C, we will always get the same result no matter how many readers / writers there are.